zrml/asset-router/src/tests/multi_currency.rs

```diff
@@ -0,0 +1,111 @@
+// Copyright 2023-2024 Forecasting Technologies LTD.
+//
+// This file is part of Zeitgeist.
+//
+// Zeitgeist is free software: you can redistribute it and/or modify it
+// under the terms of the GNU General Public License as published by the
+// Free Software Foundation, either version 3 of the License, or (at
+// your option) any later version.
+//
+// Zeitgeist is distributed in the hope that it will be useful, but
+// WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+// General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Zeitgeist. If not, see <https://www.gnu.org/licenses/>.
+
+#![cfg(test)]
+
+use super::*;
+use orml_traits::MultiCurrency;
+
+fn multicurrency_test_helper(
+    asset: Assets,
+    initial_amount: <Runtime as crate::Config>::Balance,
+    min_balance: <Runtime as crate::Config>::Balance,
+) {
+    assert_eq!(AssetRouter::minimum_balance(asset), min_balance);
+    assert_ok!(AssetRouter::deposit(asset, &ALICE, initial_amount));
+    assert_eq!(AssetRouter::total_issuance(asset), initial_amount);
+    assert_eq!(AssetRouter::total_balance(asset, &ALICE), initial_amount);
+    assert_eq!(AssetRouter::free_balance(asset, &ALICE), initial_amount);
+    assert_ok!(AssetRouter::ensure_can_withdraw(asset, &ALICE, initial_amount));
+    assert!(AssetRouter::ensure_can_withdraw(asset, &ALICE, initial_amount + 1).is_err());
+    assert_ok!(AssetRouter::transfer(asset, &ALICE, &BOB, min_balance));
+    assert_eq!(AssetRouter::free_balance(asset, &ALICE), initial_amount - min_balance);
+    assert_ok!(AssetRouter::withdraw(asset, &ALICE, 1));
+    assert_eq!(AssetRouter::free_balance(asset, &ALICE), initial_amount - min_balance - 1);
+    assert!(AssetRouter::can_slash(asset, &ALICE, 1));
+    assert_eq!(AssetRouter::slash(asset, &ALICE, 1), 0);
+    assert_eq!(AssetRouter::free_balance(asset, &ALICE), initial_amount - min_balance - 2);
+    assert_ok!(AssetRouter::update_balance(
+        asset,
+        &ALICE,
+        <AssetRouter as MultiCurrencyExtended<AccountId>>::Amount::from(1u8)
+            - <AssetRouter as MultiCurrencyExtended<AccountId>>::Amount::from(2u8)
+    ));
+    assert_eq!(AssetRouter::free_balance(asset, &ALICE), initial_amount - min_balance - 3);
```

Comment:
This is not about testing the underlying functionality in terms of completeness. The tests assume that the underlying functionality which is invoked is working properly.
The tests should only ensure that routing happens properly. If those tests are inspected in the context that they are invoked in, it makes perfect sense that proper routing is tested.

For example, for campaign assets those conditions are tested after execution of the referenced code above:
```rust
        assert_eq!(AssetRouter::total_issuance(CUSTOM_ASSET), 0);
        assert_eq!(AssetRouter::total_issuance(MARKET_ASSET), 0);
        assert_eq!(AssetRouter::total_issuance(CURRENCY), 0);
```

If those conditions are true, it must mean that all the calls above have been routed to the correct pallet, consequently routing works as expected. The only improvement that might be reasonable here from my perspective is to call `total_issuance` directly in the correct pallet instead of using the router.

Reply:
Added balance tests for Bob: https://github.com/zeitgeistpm/zeitgeist/pull/1189/commits/bcaef0e5273dffa3def7485281953ebee267a9fa

Reply:
This should improve precision of the routing tests: https://github.com/zeitgeistpm/zeitgeist/pull/1189/commits/396fed6b6c3ca22a5406388f3d69109bd395dd03 . Since it is now verified directly using the correct pallet that the total issuance is indeed correct for every pallet, the `MultiCurrency` routing tests can be considered as working properly.

Please note that the tests build upon each other, e.g. other tests assume that routing for the `MultiCurrency` trait have properly been verified.

Reply:
Also https://github.com/zeitgeistpm/zeitgeist/pull/1189/commits/f82addfe1684ed68cdce372a26f310a6f4ee7bc3